### ISSUE TITLE
Error with GPU-distributed training using --use-mas flag.

### ISSUE THREAD

**Issue by dan-ya** [dan-wells/fastpitch#1]
Thank you for re-implementation of FastPitch model!
I got an error when trying to train the model in an original FastPitch way, with `--use-mas` enabled, when using 4 GPUs.  It seems, that it crashed in the end of the 1st iteration of the 1st epoch.
I didn't have any problem, when I did a test run with CPU (at least I had 10 iterations of the 1st epoch with no problems).

**Error message**
```
-- Process 2 terminated with the following error:
Traceback (most recent call last):
  File "/PUHTI_TYKKY_8KAM643/miniconda/envs/env1/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 68, in _wrap
    fn(i, *args)
  File "fastpitch/train.py", line 662, in train
    y_pred = model(x, use_gt_durations=True)
  File "/PUHTI_TYKKY_8KAM643/miniconda/envs/env1/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1511, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/PUHTI_TYKKY_8KAM643/miniconda/envs/env1/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/PUHTI_TYKKY_8KAM643/miniconda/envs/env1/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1519, in forward
    inputs, kwargs = self._pre_forward(*inputs, **kwargs)
  File "/PUHTI_TYKKY_8KAM643/miniconda/envs/env1/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 1413, in _pre_forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by 
making sure all `forward` function outputs participate in calculating loss. 
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameter indices which did not receive grad for rank 2: 165 166
 In addition, you can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print out information about which particular parameters did not receive gradient on this rank as part of this error
```

**Log output** is the following:
```
Rank 2: Initializing distributed training
Rank 2: Done initializing distributed training
Rank 1: Initializing distributed training
Rank 1: Done initializing distributed training
Rank 3: Initializing distributed training
Rank 3: Done initializing distributed training
[2024-03-28 10:59:27] PARAMETER | output:  models
[2024-03-28 10:59:27] PARAMETER | dataset_path:  Expreso_LJS_dataset
[2024-03-28 10:59:27] PARAMETER | log_file: 
[2024-03-28 10:59:27] PARAMETER | epochs:  100
[2024-03-28 10:59:27] PARAMETER | epochs_per_checkpoint:  10
[2024-03-28 10:59:27] PARAMETER | checkpoint_path: 
[2024-03-28 10:59:27] PARAMETER | resume:  False
[2024-03-28 10:59:27] PARAMETER | seed:  1234
[2024-03-28 10:59:27] PARAMETER | amp:  False
[2024-03-28 10:59:27] PARAMETER | cuda:  True
[2024-03-28 10:59:27] PARAMETER | cudnn_benchmark:  False
[2024-03-28 10:59:27] PARAMETER | ema_decay:  0
[2024-03-28 10:59:27] PARAMETER | grad_accumulation:  1
[2024-03-28 10:59:27] PARAMETER | optimizer:  lamb
[2024-03-28 10:59:27] PARAMETER | learning_rate:  0.1
[2024-03-28 10:59:27] PARAMETER | weight_decay:  1e-06
[2024-03-28 10:59:27] PARAMETER | grad_clip_thresh:  1000.0
[2024-03-28 10:59:27] PARAMETER | batch_size:  4
[2024-03-28 10:59:27] PARAMETER | warmup_steps:  1000
[2024-03-28 10:59:27] PARAMETER | dur_predictor_loss_scale:  0.1
[2024-03-28 10:59:27] PARAMETER | pitch_predictor_loss_scale:  0.1
[2024-03-28 10:59:27] PARAMETER | attn_loss_scale:  1.0
[2024-03-28 10:59:27] PARAMETER | kl_loss_weight:  1.0
[2024-03-28 10:59:27] PARAMETER | kl_loss_start_epoch:  0
[2024-03-28 10:59:27] PARAMETER | kl_loss_warmup_epochs:  100
[2024-03-28 10:59:27] PARAMETER | training_files:  ['Expreso_LJS_dataset/expresso_fs_style.meta.train.txt']
[2024-03-28 10:59:27] PARAMETER | validation_files:  ['Expreso_LJS_dataset/expresso_fs_style.meta.val.txt']
[2024-03-28 10:59:27] PARAMETER | pitch_mean_std_file:  Expreso_LJS_dataset/pitches_stats__expresso_fs_style.json
[2024-03-28 10:59:27] PARAMETER | input_type:  char
[2024-03-28 10:59:27] PARAMETER | symbol_set:  english_basic
[2024-03-28 10:59:27] PARAMETER | text_cleaners:  []
[2024-03-28 10:59:27] PARAMETER | speaker_ids:  Expreso_LJS_dataset/speaker_ids.txt
[2024-03-28 10:59:27] PARAMETER | lang_ids:  Expreso_LJS_dataset/style_ids.txt
[2024-03-28 10:59:27] PARAMETER | hifigan:  
[2024-03-28 10:59:27] PARAMETER | hifigan_config:  hifigan/config/config_v1.json
[2024-03-28 10:59:27] PARAMETER | sampling_rate:  22050
[2024-03-28 10:59:27] PARAMETER | hop_length:  256
[2024-03-28 10:59:27] PARAMETER | audio_interval:  5
[2024-03-28 10:59:27] PARAMETER | master_addr:  localhost
[2024-03-28 10:59:27] PARAMETER | master_port:  13370
[2024-03-28 10:59:27] PARAMETER | n_mel_channels:  80
[2024-03-28 10:59:27] PARAMETER | n_symbols:  148
[2024-03-28 10:59:27] PARAMETER | padding_idx:  0
[2024-03-28 10:59:27] PARAMETER | symbols_embedding_dim:  384
[2024-03-28 10:59:27] PARAMETER | use_sepconv:  False
[2024-03-28 10:59:27] PARAMETER | use_mas:  True
[2024-03-28 10:59:27] PARAMETER | tvcgmm_k:  0
[2024-03-28 10:59:27] PARAMETER | in_fft_n_layers:  6
[2024-03-28 10:59:27] PARAMETER | in_fft_n_heads:  1
[2024-03-28 10:59:27] PARAMETER | in_fft_d_head:  64
[2024-03-28 10:59:27] PARAMETER | in_fft_conv1d_kernel_size:  3
[2024-03-28 10:59:27] PARAMETER | in_fft_conv1d_filter_size:  1536
[2024-03-28 10:59:27] PARAMETER | in_fft_sepconv:  False
[2024-03-28 10:59:27] PARAMETER | in_fft_output_size:  384
[2024-03-28 10:59:27] PARAMETER | p_in_fft_dropout:  0.1
[2024-03-28 10:59:27] PARAMETER | p_in_fft_dropatt:  0.1
[2024-03-28 10:59:27] PARAMETER | p_in_fft_dropemb:  0.0
[2024-03-28 10:59:27] PARAMETER | out_fft_n_layers:  6
[2024-03-28 10:59:27] PARAMETER | out_fft_n_heads:  1
[2024-03-28 10:59:27] PARAMETER | out_fft_d_head:  64
[2024-03-28 10:59:27] PARAMETER | out_fft_conv1d_kernel_size:  3
[2024-03-28 10:59:27] PARAMETER | out_fft_conv1d_filter_size:  1536
[2024-03-28 10:59:27] PARAMETER | out_fft_sepconv:  False
[2024-03-28 10:59:27] PARAMETER | out_fft_output_size:  384
[2024-03-28 10:59:27] PARAMETER | p_out_fft_dropout:  0.1
[2024-03-28 10:59:27] PARAMETER | p_out_fft_dropatt:  0.1
[2024-03-28 10:59:27] PARAMETER | p_out_fft_dropemb:  0.0
[2024-03-28 10:59:27] PARAMETER | dur_predictor_kernel_size:  3
[2024-03-28 10:59:27] PARAMETER | dur_predictor_filter_size:  256
[2024-03-28 10:59:27] PARAMETER | dur_predictor_sepconv:  False
[2024-03-28 10:59:27] PARAMETER | p_dur_predictor_dropout:  0.1
[2024-03-28 10:59:27] PARAMETER | dur_predictor_n_layers:  2
[2024-03-28 10:59:27] PARAMETER | pitch_predictor_kernel_size:  3
[2024-03-28 10:59:27] PARAMETER | pitch_predictor_filter_size:  256
[2024-03-28 10:59:27] PARAMETER | pitch_predictor_sepconv:  False
[2024-03-28 10:59:27] PARAMETER | p_pitch_predictor_dropout:  0.1
[2024-03-28 10:59:27] PARAMETER | pitch_predictor_n_layers:  2
[2024-03-28 10:59:27] PARAMETER | pitch_embedding_kernel_size:  3
[2024-03-28 10:59:27] PARAMETER | pitch_embedding_sepconv:  False
[2024-03-28 10:59:27] PARAMETER | speaker_cond:  ['pre']
[2024-03-28 10:59:27] PARAMETER | speaker_emb_dim:  384
[2024-03-28 10:59:27] PARAMETER | speaker_emb_weight:  1.0
[2024-03-28 10:59:27] PARAMETER | lang_cond:  ['pre']
[2024-03-28 10:59:27] PARAMETER | lang_emb_dim:  384
[2024-03-28 10:59:27] PARAMETER | lang_emb_weight:  1.0
[2024-03-28 10:59:27] PARAMETER | num_gpus:  4
[2024-03-28 10:59:27] PARAMETER | distributed_run:  True
[2024-03-28 10:59:27] PARAMETER | local_rank:  0
Rank 0: Initializing distributed training
Rank 0: Done initializing distributed training
[2024-03-28 11:01:16] epoch    1 | iter   1/2291 | loss 60.00 | mel 54.18 | dur  3.12 | pitch  1.39 | align  5.37 | took 7.27 s
```

Prior to this error (also with the current setup), I had an `AttributeError: 'float' object has no attribute 'clone'` within `train.reduce_tensor()` function with a line `rt = tensor.clone()`, when it was called from line 694:
`meta = {k: reduce_tensor(v, args.num_gpus) for k, v in meta.items()}`
The problem was that `meta['kl_weight']` was indeed a float. It was solved by introducing a line
`meta['kl_weight'] = torch.tensor(meta['kl_weight'], device=meta['loss'].get_device())`
before
`meta = {k: reduce_tensor(v, args.num_gpus) for k, v in meta.items()}.`
It could be solved in a proper way, although, I think, that my solution should not be a problem in this case.

**Comment by dan-wells** [comment]
Thanks! Your fix for the KL weight seems like the right solution to me. The error you report then was because of a missing `find_unused_parameters` option to the DDP. This should be fixed now with [473c8b6](https://github.com/dan-wells/fastpitch/commit/473c8b6d4bee3114f8f001879a0c44d6b31166c7).

P.S. You may find other things which are incomplete or not correct when using MAS or with distributed training, since I don't use either very often. Still, glad the repo is useful to you!

**Comment by dan-wells** [comment]
Actually, it was a separate issue and `find_unused_parameters` was only a workaround. There were some additional unused parameters in the MAS attention module, now removed with https://github.com/dan-wells/fastpitch/commit/566126e85385d80ee36eb7c97a033a682f9799b3.

**Comment by dan-ya** [comment]
Thank you very much for the solution!
